# Исследование надежности заемщиков


>[Исследование надежности заемщиков](#scrollTo=8a4bca55)

>>[Шаг 1. Откройте таблицу и изучите общую информацию о данных](#scrollTo=cc2556d1)

>>[Шаг 2. Предобработка данных](#scrollTo=2ce073bb)

>>>[Удаление пропусков](#scrollTo=c6dc5bb4)

>>>[Обработка аномальных значений](#scrollTo=343e4aaa)

>>>[Удаление пропусков (продолжение)](#scrollTo=fdd026fa)

>>>[Изменение типов данных](#scrollTo=eb274538)

>>>[Обработка дубликатов](#scrollTo=56510894)

>>>[Категоризация данных](#scrollTo=178164a5)

>>[Шаг 3. Исследуйте данные и ответьте на вопросы](#scrollTo=e1512691)

>>>>[3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?](#scrollTo=a7f0cf20)

>>>>[3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?](#scrollTo=a86c41fb)

>>>>[3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?](#scrollTo=6f2d99ac)

>>>>[3.4 Как разные цели кредита влияют на его возврат в срок?](#scrollTo=052eff6c)

>>>>[3.5 Приведите возможные причины появления пропусков в исходных данных.](#scrollTo=590d4639)

>>>>[3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.](#scrollTo=6d7f3b07)

>>[Шаг 4: общий вывод.](#scrollTo=65f50b41)



Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Основная цель проекта: определить как различные данные клиента банка, такие как количество детей, семейное положение, уровень дохода и разные цели кредита, влияют на его возврат в срок. Эта информация понадобится в дальнейшем для определения основных критериев, по которым будут выдаваться кредиты.

Для начала исходную таблицу нужно привести в такой вид, чтобы с ней было комфортно работать. Для этого необходимо произвести предобработку данных, чтобы затем перейти к анализу данных с целью обнаружения закономерностей, которые полностью бы отвечали на вопросы, поднятые в основной цели проекта.

## Шаг 1. Откройте таблицу и изучите общую информацию о данных

**Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [ ]:
import pandas as pd

In [ ]:
data = pd.read_csv('/datasets/data.csv')

**Выведите первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Выводы**  
В датафрейме 'data' находится информация о заёмщиках. В столбцах 'days_employed' и 'total_income' встречаются пропуски в данных. В колонке 'days_employed' есть отрицательные значения.

## Шаг 2. Предобработка данных

### Удаление пропусков

**Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [ ]:
data['education'] = data['education'].str.lower()

**Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

<ipython-input-51-0bc713f3049f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['total_income_category'] = data['total_income'].apply(categorize_income)


**Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

<ipython-input-54-ecf9d79505dd>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['purpose_category'] = data['purpose'].apply(categorize_purpose)



**Выводы**
- Мы избавились от пропусков в данных. Для колонки 'total_income' заполнили пропуски медианным значением в зависимости от типа занятости.
- Для колонки 'days_employed' изменили отрицательные значения на положительные. Удалили строки, в которых встречаются аномальные значения в столбце 'children'. Заполнили пропуски в столбце 'days_employed' медианными значениями в зависимости от типа занятости 'income_type'.
- Удалили строки с неявными дубликатами, приведя данные в столбце education к нижнему регистру.
- Создали колонку 'total_income_category' с целью категоризации данных в зависимости от уровня доходов заёмщиков.
- Создали колонку 'purpose_category' для категоризации данных о целях кредита.

## Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:
# Построим сводную таблицу, где отобразим количество кредитополучателей, количество должников и процент должников в зависимости
# от того, сколько у клиентов детей.

pivot_table_children = data.pivot_table(
    index=['children'],
    values='debt',
    aggfunc= ['count', 'sum', 'mean']
)

pivot_table_children.columns = ['Всего кредитополучателей', 'Всего должников', 'Процент должников']
pivot_table_children['Процент должников'] = pivot_table_children['Процент должников'] * 100
pivot_table_children

Всего кредитополучателей  Всего должников  Процент должников
children                                                              
0                            14091             1063           7.543822
1                             4808              444           9.234609
2                             2052              194           9.454191
3                              330               27           8.181818
4                               41                4           9.756098
5                                9                0           0.000000

**Вывод:** Для клиентов, у которых 4 и 5 детей, выборка непрезентативная. Из числа клиентов, для которых выборка является репрезентативной, процент должников меньше всего у людей, у которых нет детей. Среди клиентов, у которых 2 ребёнка, процент должников самый высокий.  

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
# Построим сводную таблицу, где отобразим количество кредитополучателей, количество должников и процент должников в зависимости
# от того, какое у клиентов семейное положение.

pivot_table_family_status = data.pivot_table(
    index='family_status',
    values='debt',
    aggfunc=['count', 'sum', 'mean']
)

pivot_table_family_status.columns = ['Всего кредитополучателей', 'Всего должников', 'Процент должников']
pivot_table_family_status['Процент должников'] = pivot_table_family_status['Процент должников'] * 100
pivot_table_family_status

Всего кредитополучателей  Всего должников  \
family_status                                                      
Не женат / не замужем                      2796              273   
в разводе                                  1189               84   
вдовец / вдова                              951               63   
гражданский брак                           4134              385   
женат / замужем                           12261              927   

                       Процент должников  
family_status                             
Не женат / не замужем           9.763948  
в разводе                       7.064760  
вдовец / вдова                  6.624606  
гражданский брак                9.313014  
женат / замужем                 7.560558

**Вывод:** Вдовцы / вдовы имеют наибольшее отношение отсутствия задолженности по возврату кредитов к общему количеству выданных кредитов. Те, кто не женаты/ не замужем и клиенты, находящиеся в гражданском браке, имеют наименьший показатель отсутствия задолженности по возврату кредитов к общему количеству выданных кредитов.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
# Построим сводную таблицу, где отобразим количество кредитополучателей, количество должников и процент должников в зависимости
# от того, к какой категории по доходам относится клиент.

pivot_table_total_income_category = data.pivot_table(index='total_income_category',
                                                    values='debt',
                                                    aggfunc=['count', 'sum', 'mean']
)
pivot_table_total_income_category.columns = ['Всего кредитополучателей', 'Всего должников', 'Процент должников']
pivot_table_total_income_category['Процент должников'] = pivot_table_total_income_category['Процент должников'] * 100
pivot_table_total_income_category

Всего кредитополучателей  Всего должников  \
total_income_category                                              
A                                            25                2   
B                                          5014              354   
C                                         15921             1353   
D                                           349               21   
E                                            22                2   

                       Процент должников  
total_income_category                     
A                               8.000000  
B                               7.060231  
C                               8.498210  
D                               6.017192  
E                               9.090909

**Вывод:** Для клиентов, у которых категория дохода 'A' и 'E', выборка непрезентативная. Из числа клиентов, для которых выборка является репрезентативной, процент должников меньше всего у людей, у которых категория доходов D . Для клиентов, у которых категория доходов C, процент должников самый высокий.  
Таким образом, если сравнить самые большие категории B и C, можно сказать, что заемщики с более высоким доходом платят кредиты вовремя лучше.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
# Построим сводную таблицу, где отобразим количество кредитополучателей, количество должников и процент должников в зависимости
# от целей кредита.

pivot_table_purpose_category = data.pivot_table(index='purpose_category',
                                                    values='debt',
                                                    aggfunc=['count', 'sum', 'mean']
)
pivot_table_purpose_category.columns = ['Всего кредитополучателей', 'Всего должников', 'Процент должников']
pivot_table_purpose_category['Процент должников'] = pivot_table_purpose_category['Процент должников'] * 100
pivot_table_purpose_category

Всего кредитополучателей  Всего должников  \
purpose_category                                                      
операции с автомобилем                        4279              400   
операции с недвижимостью                     10751              780   
получение образования                         3988              369   
проведение свадьбы                            2313              183   

                          Процент должников  
purpose_category                             
операции с автомобилем             9.347978  
операции с недвижимостью           7.255139  
получение образования              9.252758  
проведение свадьбы                 7.911803

**Вывод:** Для целей кредита "операции с недвижимостью" и "проведение свадьбы" чаще всего отсутствует задолженность. Отношение отсутствия задолженности к числу выданных кредитов составляят 0.927449 и 0.920882 для двух данных категорий, соответственно.
    Для целей кредита "операции с автомобилем " и "получение образования" чаще присутствует задолженность. Отношение отсутствия задолженности к числу выданных кредитов составляет 0.906520 и 0.907472 для двух данных категорий, соответственно.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Количество пропущенных значений в столбцах 'days_employed' и 'total_income' одинаково - 2174. Из-за этого можно предположить, что возможной причиной появления пропусков в исходных данных является техническая причина.  
Кроме технических ошибок причиной возникновения пропусков в данных может быть человеческий фактор.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Заполнение пропусков медианным значением для количественных переменных является лучшим решением, так как оторванные  от основной массы значения смещают средний показатель вверх. Медиана менее чувствительна к выбросам.  
Для примера рассчитаем эти метрики для столбца `days_employed`, где есть выбросы.

In [ ]:
print(f'Среднее значение столбца "days_employed": {data["days_employed"].mean():.2f}')
print(f'Медианное значение столбца "days_employed": {data["days_employed"].median():.2f}')

Среднее значение столбца "days_employed": 67148.97
Медианное значение столбца "days_employed": 1996.73


## Шаг 4: общий вывод.

По итогу исследования можно сделать следующие выводы:

- Из числа клиентов по количеству детей, для которых выборка является репрезентативной, процент должников меньше всего у людей, у которых нет детей. Среди клиентов, у которых 2 ребёнка, процент должников самый высокий. Для клиентов, у которых есть 4 или 5 ребёнка, выборка непрезентативная.
- Состоящие или состоявшие в браке заемщики реже допускают просрочки.
- Процент должников наименьший среди клиентов с категорией доходов 'D' (30001–50000). Самый высокий процент должников среди клиентов категории 'C' (50001–200000). Выборка оказалась нерепрезентативной для клиентов с категорие доходов 'A' (1000001 и выше) и 'E' (0–30000).
- Для целей кредита "операции с недвижимостью" и "проведение свадьбы" чаще всего отсутствует задолженность по возврату кредита. Для целей кредита "операции с автомобилем " и "получение образования" задолженность присутствует чаще, чем для остальных целей.